carbon-design-system/carbon
# feat(upgrade): run migrations from upgrade cli

Closes #10466

This PR updates the `@carbon/upgrade` CLI to be able to run one or more `migrations` as part of an upgrade flow. These `migrations` contain a `migrate` function that can run jscodeshift codemods.

The following is now possible:
- `@carbon/upgrade`: upgrade and run all available migrations for a given upgrade
- ~`@carbon/upgrade migrate`: prompt a list of all available migrations, select one to run~ removed based on feedback
- `@carbon/upgrade migrate list`: list all available migrations
- `@carbon/upgrade migrate <migration-name>`: run a specific migration

#### Changelog

**New**

- Provide new `migrate` command to run and list individual migrations
- ~Add a new `imports-to-unified-package.js` codemod to rewrite imports from `carbon-components-react` to `@carbon/react`~ moved to https://github.com/carbon-design-system/carbon/pull/10523

**Changed**

- Modify the upgrade flow to run all available migrations for a given upgrade flow by default
- Move `transforms` from `codemods` package to live within the `upgrade` package

**Removed**

- Remove `codemods` package

#### Testing / Reviewing

- Try out the CLI locally by:
  - Pulling down this PR
  - Run `cd packages/upgrade`
  - Run `yarn build` to build the CLI
  - Run `cd fixtures/sample-project` to use our sample project fixture
  - Run `../../bin/carbon-upgrade.js --help` to verify the CLI is up-and-running
  - Take a look at the options / capabilities of what's there to see if it makes sense
  - Run `../../bin/carbon-upgrade.js` and choose an upgrade path. Verify the CLI runs all the available migrations for that upgrade path. In the case of `v11: carbon-components`, two stub migrations should be called and log out to the console.
  - Run `../../bin/carbon-upgrade.js migrate list` to see a list of migrations should be shown
  - ~Run `../../bin/carbon-upgrade.js migrate` to see a prompt of migrations should be shown allowing a single selection~ removed based on feedback
